EOS-20758: Implementation of API using current S3 implementation as reference includes MD5 - client side

gshipra opened the pull request:
Implementation of API using current S3
implementation as reference includes MD5 - client side

Problem:
Do the implementation for motr client api,
implementation include calculations for MD5 checksums and digest
and the test code.

Solution:
Done the implementation for calculating MD5checksums and digest
Written code for UT.

Testing
Code compiled and newly written UT passed.

Signed-off-by: Shipra Gupta <shipra.gupta@seagate.com>

gshipra committed:
EOS-20758: Implementation of API using current S3
implementation as reference includes MD5 - client side

Problem:
Do the implemnetation for motr client api,
implementation include calculations for MD5 checksums and digest
and the test code.

Solution:
Done the implementation for calculating MD5checksums and digest
Written code for UT.

Testing
Code compiled and newly written UT passed.

Signed-off-by: Shipra Gupta <shipra.gupta@seagate.com>

sg-shankar commented on motr/client.c:
```diff
@@ -897,23 +905,41 @@ M0_INTERNAL int m0_calculate_md5_inc_digest(
+			M0_LOG(M0_ERROR, "MD5_Final failed with error %d", rc);
+			return M0_ERR(rc);
+		}
 	}
```
For KV and case where user wants PI value without seed, this else part will be needed

sg-shankar commented on motr/client.c:
```diff
@@ -897,23 +905,41 @@ M0_INTERNAL int m0_calculate_md5_inc_digest(
 	 */
 	if (seed) {
+
+		char seed_str[128];
```
This can be seed_str[sizeof(struct m0_pi_seed)] with round off? or add comments regarding why 128 was used

sg-shankar commented on motr/ut/protection_info_checks.c:
```diff
@@ -0,0 +1,221 @@
+	m0_fid_set(&seed.obj_id, 0x123, 0x456);
+
+	memset(&pi, 0, sizeof(struct m0_md5_inc_digest_pi));
+	pi.hdr.pi_type = 1;
```
For readability you can add md5_with_digest enum

sg-shankar commented on motr/ut/protection_info_checks.c:
```diff
@@ -0,0 +1,221 @@
+		}
+	}
+
+	m0_fid_set(&seed.obj_id, 0x123, 0x456);
```
#define these hard coded value and use it at both place

sg-shankar reviewed (commented):
Please add test and validation for following scenario:

_Scenario 1: Validate that calling API once or multiple time gives same value of pi_value_without_seed_

e.g Pass 64k x 10 data unit in one go to API with no seed and compute pi_without_seed. This value should match pi_without_seed comuted at the end of calling API each times for 10 data unit

_Scenario 2: Calling API many time for a Data Unit for each chunk or segment of Data Unit and calling once give same result for seeded PI value._

DU having many buffers e.g. 10 buffer : e.g. Call API 10 time, with no seed for 9 call in the 10th call pass some seed which is some non-zero offset; similarly now call API one time with all buffers-chunk with same seed, the output should match

gshipra committed:
EOS-20758: Implementation of API using current S3
implementation as reference includes MD5 - client side

Problem:
Do the implemnetation for motr client api,
implementation include calculations for MD5 checksums and digest
and the test code.

Solution:
Done the implementation for calculating MD5checksums and digest
Written code for UT.

Testing
Code compiled and newly written UT passed.

Signed-off-by: Shipra Gupta <shipra.gupta@seagate.com>

sg-shankar commented on motr/client.c:
```diff
@@ -859,7 +859,7 @@ M0_INTERNAL int m0_calculate_md5_inc_digest(
 #define PI_DIGEST_LENGTH 92
```
This part of code can now be removed

sg-shankar commented on motr/client.c:
```diff
@@ -897,23 +905,54 @@ M0_INTERNAL int m0_calculate_md5_inc_digest(
+		}
 	}
 
+	rc = MD5_Final(pi->pi_value, &digest);
```
Lets add flag where user can skip computation of this PI Value for optimization.

sg-shankar commented on motr/ut/protection_info_checks.c:
```diff
@@ -0,0 +1,336 @@
+			M0_PI_CALC_UNIT_ZERO, big_curr_digest, NULL);
+	M0_ASSERT(rc == 0);
+
+	M0_ASSERT(memcmp(&seeded_final_chunks_value, pi.pi_value, MD5_DIGEST_LENGTH) == 0);
```
Lets add validation for pi value without seed also
